hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -1114,19 +1155,59 @@ def uniform(random, n):
     return int_to_bytes(random.getrandbits(n * 8), n)
 
 
+def pop_random(random, values):
+    """Remove a random element of values, possibly changing the ordering of its
+    elements."""
+
+    # We pick the element at a random index. Rather than removing that element
+    # from the list (which would be an O(n) operation), we pop the last element
+    # from the list and overwrite the chosen index. This changes the order of
+    # the elements, but as long as these elements are only accessed through
+    # random sampling that doesn't matter.
+    i = random.randrange(0, len(values))
+    result = values[i]
+    end = values.pop()
+
+    # If i was already the last index in the list then there's no need to
+    # overwrite the value there.
+    if i < len(values):
+        values[i] = end
+    return result
```

Comment:
Consider replacing this implementation with:

    i = random.randrange(0, len(values))
    values[i], values[-1] = values[-1], values[i]
    return values.pop()

This makes the "swap and pop" trick a little easier to see. And because it updates `values[i]` before popping, it doesn't need a special case when choosing the last element.

Reply:
FYI, you can get syntax highlighting in github markdown by surrounding the block in triple-backticks, and tagging it with the syntax, eg

    ```python
    i = random.randrange(0, len(values))
    ```

(without `python` it's a generic monospaced block)